algoliasearch/algoliasearch.go

```diff
@@ -313,6 +313,44 @@ type Index interface {
 	// is only kept for backward-compatibility reason as we decided to change
 	// its name.
 	SearchFacet(facet, query string, params Map) (res SearchFacetRes, err error)
+
+	// SaveRule saves the given Rule for the current index. If a Rule with the
+	// same objectID already exists, it will get overriden. The operation can
+	// be forwarded to the index replicas by setting `forwardToReplicas` to
+	// `true`.
+	SaveRule(rule Rule, forwardToReplicas bool) (SaveRuleRes, error)
+
+	// SaveRule saves the given Rules by batch for the current index. The
+	// operation can be forwarded to the index replicas by setting
+	// `forwardToReplicas` to `true`. A `clear` operation can be applied before
+	// writing the new Rules by setting `clearExistingRules` to `true`.
+	BatchRules(rules []Rule, forwardToReplicas, clearExistingRules bool) (BatchRulesRes, error)
+
+	// GetRule returns the Rule identified by the given `objectID`. A non-nil
+	// error is returned if the Rule cannot be found.
+	GetRule(objectID string) (*Rule, error)
+
+	// DeleteRule deletes the Rule identified by the given `objectID` for the
+	// current index. The operation can be forwarded to the index replicas by
+	// setting `forwardToReplicas` to `true`.
+	DeleteRule(objectID string, forwardToReplicas bool) (DeleteRuleRes, error)
+
+	// ClearRules removes all existing Rules for the current index. The
+	// operation can be forwarded to the index replicas by setting
+	// `forwardToReplicas` to `true`.
+	ClearRules(forwardToReplicas bool) (ClearRulesRes, error)
+
+	// SearchForRules allows to search for Rules for the current index. All
+	// arguments are optional, using an empty string for string and
+	// RulePatternAnchoring parameters and `-1` for integer parameters. For
+	// more details regarding the parameters:
+	//
+	//   - `query`:       enable full text search within the Rule fields
+	//   - `anchoring`:   restricts the search to Rules with a specific anchoring type
+	//   - `context`:     restricts the search to rules with a specific context (exact match)
+	//   - `page`:        requested page (zero-based, defaults to zero)
+	//   - `hitsPerPage`: maximum number of hits in a page (defaults to 20)
+	SearchForRules(query string, anchoring RulePatternAnchoring, context string, page, hitsPerPage int) (SearchForRulesRes, error)
```

Comment:
I didn't know it was subject to modifications. It's already painful to pass 5 parameters. I'll make it take a map and perform a type check on the known parameters. This way, even if a new parameter is used by the user, it will let it go through.